[Reporting/Dashboard] Use Chromium for print-optimized PDFs

jloleysens opened the pull request:
## Summary

It is possible to request a "print-optimized" PDF from dashboard (our most popular app for making print-focussed reports). Today we use a native NPM dependency instead of Chromium for generating the PDF. We can improve performance of generating these PDFs by leveraging the print functionality native to browsers since we are already using headless Chromium.

## Screenshots

Note: header and footers are included via puppeteers `footerTemplate` and `headerTemplate` (see [their docs](https://puppeteer.github.io/puppeteer/docs/9.1.1/puppeteer.pdfoptions.headertemplate)). Note: header is outdated in screenshot

<img width="745" alt="Screenshot 2022-04-20 at 18 54 10" src="https://user-images.githubusercontent.com/8155004/164284006-6ab40df2-2d69-4fb5-a205-aeed5f43fb3d.png">

## Samples

* (Linux) [[Logs] Web Traffic 3 (1).pdf](https://github.com/elastic/kibana/files/8630051/Logs.Web.Traffic.3.1.pdf)
* (Linux) [[Logs] Web Traffic 3 (2) with custom logo.pdf](https://github.com/elastic/kibana/files/8630058/Logs.Web.Traffic.3.2.with.custom.logo.pdf)




## Benchmarks

See https://github.com/elastic/kibana/pull/130546#issuecomment-1103208779

### Additional notes

* Rework of https://github.com/elastic/kibana/pull/128470
* ~Needs benchmarks (CC @tsullivan )~
* Current implementation is server-side only, even though this does get us a lot closer to being able to printing client side, we still need to figure out how we want to handle headers/footers there.
* Fix for missing time range in print-optimized PDF https://github.com/elastic/kibana/pull/130546/commits/c3b5500353960f7f5b36456d785055cef9909ea0


### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios


### Risk Matrix


| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Introduce a performance regression. | Low | High | Running benchmarks will help a lot with this. We are also taking fewer logical steps to create a PDF and not cloning/stringifying image buffers which should already greatly increase performance. |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

jloleysens commented:
@elasticmachine merge upstream

tsullivan commented:
## Benchmark tests

These results used the [three "print layout" tests](https://github.com/elastic/kibana-reporting-benchmarks/blob/main/tests/run_tests.sh#L18-L20) from elastic/kibana-reporting-benchmarks repo. 

The average execution time using the changes in this PR: **12.846 seconds**
The average execution time baseline from the latest 8.3.0-SNAPSHOT build: **17.241 seconds**.

Take a look at the dashboard set up from the logs of the test runs. The "Execution duration" chart is split by the server UUID: the blue line is the timings with a build of this PR, and the green line is the snapshot build.
![image](https://user-images.githubusercontent.com/908371/164107894-6a7246f4-9fb2-49bd-84fe-2f067bc51d24.png)

The other stats: event loop delay, heap memory size, chromium memory size, etc. look virtually the same.

jloleysens commented:
If the styles of footers and headers is an issue we can consider using something like this: https://www.npmjs.com/package/puppeteer-report

jloleysens commented on x-pack/test/reporting_api_integration/reporting_and_security/usage/_post_urls.ts:
```diff
@@ -19,7 +19,7 @@ export const PDF_PRESERVE_PIE_VISUALIZATION_6_3 = `/api/reporting/generate/print
 
 export const PDF_PRINT_PIE_VISUALIZATION_FILTER_AND_SAVED_SEARCH_6_3 = `/api/reporting/generate/printablePdf?jobParams=${encodeURIComponent(
-  `(browserTimezone:America/New_York,layout:(id:print),objectType:visualization,relativeUrls:!('/app/kibana#/visualize/edit/befdb6b0-3e59-11e8-9fc3-39e49624228e?_g=(refreshInterval:(display:Off,pause:!!f,value:0),time:(from:!'Mon+Apr+09+2018+17:56:08+GMT-0400!',mode:absolute,to:!'Wed+Apr+11+2018+17:56:08+GMT-0400!'))&_a=(filters:!!((!'$state!':(store:appState),meta:(alias:!!n,disabled:!!f,index:a ...
+  `(browserTimezone:America/New_York,layout:(dimensions:(height:588,width:1038),id:preserve_layout),objectType:visualization,relativeUrls:!('/app/kibana#/visualize/edit/befdb6b0-3e59-11e8-9fc3-39e49624228e?_g=(refreshInterval:(display:Off,pause:!!f,value:0),time:(from:!'Mon+Apr+09+2018+17:56:08+GMT-0400!',mode:absolute,to:!'Wed+Apr+11+2018+17:56:08+GMT-0400!'))&_a=(filters:!!((!'$state!':(store:a ...
```
I had to update this because it was breaking an old POST URL request. This could be an issue if people are using old URLs for visualize that have a `printablePdf` using a `print` layout.


Broken:
[Filter Test_ animals_ linked to search with filter.pdf](https://github.com/elastic/kibana/files/8528987/Filter.Test_.animals_.linked.to.search.with.filter.pdf)

Fixed:
[Filter Test_ animals_ linked to search with filter (1).pdf](https://github.com/elastic/kibana/files/8528994/Filter.Test_.animals_.linked.to.search.with.filter.1.pdf)

@tsullivan should we consider this a breaking change? I can't see how users could request a report like this from us today given that we don't supply print layout reports from visualize.

tsullivan replied:
> I can't see how users could request a report like this from us today given that we don't supply print layout reports from visualize.

Agreed. It's weird the test was for an export type that isn't supported in the UI. I don't consider this a breaking change, as this seems to have been an "internal-only" export type.

elasticmachine commented:
Pinging @elastic/kibana-reporting-services (Team:Reporting Services)

elasticmachine commented:
Pinging @elastic/kibana-app-services (Team:AppServicesUx)

jloleysens commented on x-pack/plugins/screenshotting/server/formats/pdf/index.ts:
```diff
@@ -6,6 +6,9 @@
 import { groupBy } from 'lodash';
+// FIXME: Once/if we have the ability to get page count directly from Chrome/puppeteer
+// we should get rid of this lib.
+import * as PDFJS from 'pdfjs-dist/legacy/build/pdf.js';
```
I could not see another way to reliably get page numbers from the PDF buffer produced by Chromium. The puppeteer API does not allow for this and I could not figure out how to do this with `pdfmake`.

`pdfjs` is 100% JS so fortunately no extra native deps, but it does add a fairly large module for a singular purpose - getting page numbers.

tsullivan replied:
I put some time into this as well. I hoped it would be easy to find the page count somehow by parsing a PDF buffer, but no luck.

There may be a way to guess the number of pages, based on the number of visualization containers on the page, but we don't pass that info down to this scope.

jloleysens commented:
@elasticmachine merge upstream

mdefazio reviewed (approved):
Styles look good. 

Do we need to update _vars for other page sizes? 

Maybe i'm misinterpreting the usage here, but looking at [page-size](https://drafts.csswg.org/css-page-3/#typedef-page-size-page-size), can we simply put in the string of the supported format to get the sizes? 

This already looks far better than before, so i'm not suggesting a change here, but curious if we need to consider this in the future. Maybe just supporting 'letter'?

jloleysens commented:
Thanks for the review @mdefazio !

> Do we need to update _vars for other page sizes?

Yeah, we will need to update this. The current implementation is fixed to portrait A4 print outs, since that is all we support. I also exposed the "content height" to dashboard by doing a bit of math (content = height - header - footer) so that dashboard knows how big it can make visuals to fill up a page.

In future I'd like to get coverage for different page sizes, but need to do a bit more research TBH. We might be able to do something like `100vh - headerHeight - footerHeight` if we can get those vars to change dynamically based on a media query 🤔

tsullivan commented on src/plugins/dashboard/public/application/dashboard_app.tsx:
```diff
@@ -119,13 +119,12 @@ export function DashboardApp({
+            redirectTo={redirectTo}
+            embedSettings={embedSettings}
+            dashboardAppState={dashboardAppState}
+          />
```
does this affect preserve layout?

tsullivan commented on x-pack/plugins/screenshotting/server/browsers/chromium/templates/footer.html:
```diff
@@ -0,0 +1,37 @@
+    width: 20mm;
+
+    content: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAkEAAADGCAMAAADi8KTWAAAA9lBMVEUAAAB9fn6FhoaMjY2MjY2EhISIiYlubGyFhoZlZWWMjY2MjY2LjIyEhYWMjY2Ki4t9fn6HiIiNjo6IiYmJiYmLjIyBgYF+f3+GiYmLjIyIiYmHiIiBgoKPkJCHiIiGh4eMjY2JioqMjY2LjIyNjo6MjY2FhoaIioqKi4uJioqKi4uFhoaIioqHiIiFhoaLjIyKjIyKi4uJioqOj4+Njo6MjY2LjIx8fX2Njo6LjIyOj4+Oj4+Njo6Ki4uMjY2LjIyKi4uLjIyPkJCLjIyNj4+MjY2LjIyOjo56e3uLjIyKi4uIiYm ...
+  }
```
user's custom logo?

jloleysens replied:
I was thinking of that as a future enhancement. Perhaps we could change this to a handlebars template to make those values dynamic.

tsullivan commented on x-pack/plugins/screenshotting/server/formats/pdf/pdf_maker/pdfmaker.ts:
```diff
@@ -83,6 +83,8 @@ export class PdfMaker {
     const groupCount = this.content.length;
 
     // inject a page break for every 2 groups on the page
+    // TODO: Remove this code since we are now using Chromium to drive this
```
Is this planned to do in this PR? It seems reasonable to.

jloleysens replied:
I'd like to do that in a separate PR mainly because I've been trying to focussing changes on print layout reports where that code path also touches other reports and not changing it now does not affect those.

Does that sound OK to you?

tsullivan commented:
I see some minor rendering issues:
- Maps controls showing:
  ![image](https://user-images.githubusercontent.com/908371/164558827-b5ef3662-6750-4b59-855a-1b46e6527009.png)
- Controls showing in a saved search panel:
  ![image](https://user-images.githubusercontent.com/908371/164559102-46aebd77-1359-45db-980a-bc5e178c7b53.png)

Update: These are unrelated: after checking the `main` branch, I found the issues are there as well :slightly_frowning_face:

tsullivan reviewed (commented):
This change is really fantastic! Overall, the code changes look great. I tested things out and only found one issue: that page number seems to not be working.

jloleysens commented on x-pack/plugins/screenshotting/server/browsers/chromium/templates/footer.html:
```diff
@@ -0,0 +1,37 @@
+<div class="container">
+  <div class="logo"></div>
+  <div class="pages">
+    <span class="pageNumber"></span>&nbsp;of&nbsp;<span class="totalPages"></span>
```
Hm, it looks like the version of Chromium being used doesn't interpolate the page number values per the docs https://puppeteer.github.io/puppeteer/docs/9.1.1/puppeteer.pdfoptions/. I can see this is an issue on the Cloud version too. That is really unfortunate, we might need to test with a newer version of Chromium.

jloleysens replied:
It works on the one we use on macOS FWIW.

jloleysens replied:
TLDR; looks like there is a bug on linux-based Chromium and rendering page numbers.

I looked at this a bit closer and it looks like Chromium falls back to using this HTML for the header and footer if we do not pass in a value (based on our Chromium revision):

https://source.chromium.org/chromium/chromium/src/+/refs/tags/93.0.4577.0:components/printing/resources/print_header_footer_template_page.html;bpv=0;bpt=0

Testing without passing through a value renders their default that ends up looking bugged:

<img width="949" alt="Screenshot 2022-04-22 at 15 27 30" src="https://user-images.githubusercontent.com/8155004/164723638-5254cba9-efe5-460e-adb1-a79c600a163a.png">

I think we should first try to upgrade puppeteer (and Chromium). Hopefully this is fixed on a newer version.

jloleysens commented:
@elasticmachine merge upstream

jloleysens commented:
@elasticmachine merge upstream

jloleysens commented:
@elasticmachine merge upstream

jloleysens commented:
@elasticmachine merge upstream

nreese reviewed (approved):
maps changes LGTM
code review

jloleysens commented:
@elasticmachine merge upstream

jloleysens commented:
@elasticmachine merge upstream

jloleysens commented:
@tsullivan I think I've addressed your feedback:

* Enable custom logos (see attached PDF in description)
* Added the `headless_shell/chromium-70f5d88-locales-linux_arm64.zip` and `headless_shell/chromium-70f5d88-locales-linux_x64.zip` archives that contain the `locales` we need for the page numbers

we should also merge https://github.com/elastic/kibana/pull/131600 but I don't think we need to block this PR.

tsullivan commented on x-pack/test/reporting_api_integration/services/usage.ts:
```diff
@@ -101,31 +101,47 @@ export function createUsageServices({ getService }: FtrProviderContext) {
+      log.info(
+        `expecting all time stats to report ${count} ${jobType} job types (actual: ${actual})`
+      );
+      expect(actual).to.be(count);
```
Sorry this test was such a pain. It's very old and was made before snapshots were a thing.

tsullivan reviewed (approved):
LGTM

I have some percolating thoughts about calculating the page numbers of the PDF for telemetry, but let's save those for later.

spalger reviewed (approved):
Casing changes LGTM

Heenawter reviewed (approved):
Code owner review - LGTM 👍

jloleysens commented:
@elasticmachine merge upstream
